### PR TITLE
Make training_epoch_end behave like validation_epoch_end

### LINKED ISSUE #914: Log training metrics for each epoch
Currently, I am able to log training metrics to Tensorboard using:

```
import pytorch_lightning as pl
from pytorch_lightning.loggers import TensorBoardLogger

logger = TensorBoardLogger(save_dir=save_dir, name="my_model")

[...]

trainer = pl.Trainer(logger=logger)
```

This logs training metrics (loss, for instance) after each batch. I would like to be able to average these metrics across all batches and log them to TensorBoard only once, at the end of each epoch. This is what the `validation_end` method does in your example: https://github.com/PyTorchLightning/pytorch-lightning/blob/446a1e23d7fe3b2e07f1a5887fe819d0dfa7d4e0/pl_examples/basic_examples/lightning_module_template.py#L145. 

I first thought about writing my own `training_end` method. But this method is called after each batch instead of being called at the end of an epoch (as I would have thought). The method `on_epoch_end` seems interesting but does not receive an `outputs` argument as `training_end` does. Basically, in my model, I would like to write something like:  `self.logger.experiment.add_scalar('training_loss', train_loss_mean, global_step=self.current_epoch)`, but I do not know where to put this line.

 - OS: Debian GNU/Linux 9.11 (stretch)
 - Packaging: PIP
 - Version 0.6.1.dev0

### PR DESCRIPTION
**Pull request by jbschiratti**
## What does this PR do?
This PR fixes #914. 

The class [`LightningModule`](https://github.com/PyTorchLightning/pytorch-lightning/blob/1576ad9963402c1a6f6353df89386d8fe66f5dac/pytorch_lightning/core/lightning.py#L32) currently implements a [`validation_epoch_end`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/lightning.py#L451) method but no `training_epoch_end` method. The [`CHANGELOG.md`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md) file suggests that, from version 0.7.0, the `training_end` (or now [`training_step_end`](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/lightning.py#L234)) method is to be renamed `training_epoch_end`. However, this is not satisfactory as `training_epoch_end` would still not behave like `validation_epoch_end`. 

This PR ensures that both `training_epoch_end` and `validation_epoch_end` have the same behavior (for instance: allow the user to log metrics only once at the end of each epoch). 

## PR review    
Anyone in the community is free to review the PR once the tests have passed 👍

### CODE REVIEW

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -385,6 +386,9 @@ def train(self):
 
     def run_training_epoch(self):
 
+        # get model
+        model = self.get_model()
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:390]: with this, you can remove/replace all subsequent `self.get_model()` calls

File: pytorch_lightning/trainer/training_loop.py [hunk 2]
```diff
@@ -483,6 +491,17 @@ def run_training_epoch(self):
...
                 break
 
+        # process epoch outputs
+        if isinstance(model, (LightningDistributedDataParallel, LightningDataParallel)):
+            model = model.module
+
+        if self.is_overriden('training_epoch_end', model=model):
+            epoch_output = model.training_epoch_end(outputs)
+            _, _, log_epoch_metrics, callback_epoch_metrics, _ = self.process_output(
+                epoch_output)
```
Comment (Borda) [inline pytorch_lightning/trainer/training_loop.py:501]:
```suggestion
            _, _, log_epoch_metrics, callback_epoch_metrics, _ = self.process_output(epoch_output)
```

Comment (jbschiratti) [reply pytorch_lightning/trainer/training_loop.py:501]: done! I wrote it a bit differently to avoid having a line break character `\` after `=`.

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -232,6 +232,78 @@ def training_end(self, *args, **kwargs):
...
+        Args:
+            outputs: List of outputs you defined in training_step, or if there are multiple
+            dataloaders, a list containing a list of outputs for each dataloader
+
+        Return:
+            Dict or OrderedDict (dict): Dict has the following optional keys:
+            progress_bar -> Dict for progress bar display. Must have only tensors
+            log -> Dict of metrics to add to logger. Must have only tensors (no images, etc)
+
+        .. note:: If you this method is not overridden, this won't be called.
```
Comment (ethanwharris) [inline pytorch_lightning/core/lightning.py:260]: Typo -> If this method is not ...

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -545,14 +564,13 @@ def run_training_batch(self, batch, batch_idx):
...
+                        output_dict = self.training_forward(
                             split_batch, batch_idx, opt_idx, self.hiddens)
 
-                    closure_loss = output[0]
-                    progress_bar_metrics = output[1]
-                    log_metrics = output[2]
-                    callback_metrics = output[3]
-                    self.hiddens = output[4]
+                        # format and reduce outputs accordingly
+                        processed_output = self.process_output(output_dict, train=True)
```
Comment (ethanwharris) [inline pytorch_lightning/trainer/training_loop.py:571]: This doesn't detach the members of the output_dict, so gradient trees will remain in memory

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (jbschiratti) [reply pytorch_lightning/core/lightning.py]: Fixed, thanks.

File: pytorch_lightning/trainer/training_loop.py (hunk 4 above)
Comment (jbschiratti) [reply pytorch_lightning/trainer/training_loop.py:573]: You're absolutely right ! Thank you for pointing this out.
Have a look at the [latest commit](https://github.com/PyTorchLightning/pytorch-lightning/pull/1357/commits/39cecdba278bf8143bd746b934a7198d98d0237a) and tell me if this fix is OK with you.

Comment (ethanwharris) [reply pytorch_lightning/trainer/training_loop.py:573]: Awesome, that fixes the problem AFAICT, nice work! :)

**Comment by jbschiratti** [comment]
CircleCI failed on `test_ddp_all_dataloaders_passed_to_fit` however, AFAIK, this does not seem to be related with the PR. Is it?

**Comment by jbschiratti** [comment]
Thanks @Borda, @ethanwharris and @justusschock 👍🎉!

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -231,6 +231,78 @@ def training_end(self, *args, **kwargs):
...
+
+            train_outs = []
+            for train_batch in train_data:
+                out = training_step(train_batch)
+                train_outs.append(out)
+            training_epoch_end(val_outs)
+
+        Args:
+            outputs: List of outputs you defined in training_step, or if there are multiple
+            dataloaders, a list containing a list of outputs for each dataloader
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:252]: there should an indent here, otherwise it messes up the formatting. Please remember to always generate the docs and inspect them when updating the docs. Just because there are no build errors doesn't mean it looks as intended.

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -231,6 +231,78 @@ def training_end(self, *args, **kwargs):
...
+
+                    # log training accuracy at the end of an epoch
+                    results = {
+                        'log': {'train_acc': train_acc_mean.item()}
+                    }
+                    return results
+
+            With multiple dataloaders, `outputs` will be a list of lists. The outer list contains
+            one entry per dataloader, while the inner list contains the individual outputs of
+            each validation step for that dataloader.
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:285]: it should say training step

**Comment by Borda** [comment]
@awaelchli thank you for your careful reading, may apply your suggestions in follow-up PR?

**Comment by awaelchli** [comment]
yeah sure:)
btw is it not possible to add more commits to this branch and merge again?

**Comment by Borda** [comment]
well, you would need to fork `jbschiratti:fix_914` (another forked repo) so I would suggest starting with the actual master as this is already merged...

**Comment by jbschiratti** [comment]
@Borda I can push new commits to this branch and you may be able to merge it again into master :-)
